docs/guides/ssr.md

```diff
@@ -22,9 +20,214 @@ The most basic server rendering in React Router is pretty straightforward. Howev
+    init.body = req.body;
+  }
+
+  return new Request(url.href, init);
+};
+```
+
+Once we've loaded our data by executing all of the matched route loaders for the incoming request, we use `createStaticRouter()` and `<StaticRouterProvider>` to render the HTML and send a response back to the browser:
+
+```js filename=server.jsx lines=[5-16]
+app.get("*", async (req, res) => {
+  let fetchRequest = createFetchRequest(req);
+  let context = await handler.query(fetchRequest);
+
+  let router = createStaticRouter(
+    handler.dataRoutes,
+    context
+  );
+  let html = ReactDOMServer.renderToString(
+    <StaticRouterProvider
+      router={router}
+      context={context}
+    />
+  );
+
+  res.send("<!DOCTYPE html>" + html);
+});
+```
+
+Once we've sent the HTML back to the browser, we'll need to "hydrate" the application on the client using `createBrowserRouter()` and `<RouterProvider>`:
+
+```jsx filename=entry-client.jsx lines=[10-15]
+import * as React from "react";
+import ReactDOM from "react-dom/client";
+import {
+  createBrowserRouter,
+  RouterProvider,
+} from "react-router-dom";
+
+import { routes } from "./routes";
+
+let router = createBrowserRouter(routes);
+
+ReactDOM.hydrateRoot(
+  document.getElementById("app"),
+  <RouterProvider router={router} />
+);
+```
+
+And with that you've got a server-side-rendered and hydrated application! For a working example, you may also refer to the [example][ssr-data-router-example] in the Github repository.
+
+### Additional Concepts
+
+As mentioned above, server-side rendering is tricky at scale and for production-grade applications, and we strongly recommend checking out [Remix][remix] if that's your goal. But if you are going the manual route, here's a few additional concepts you may need to consider:
+
+#### Redirects
+
+If any loaders redirect, `handler.query` will return the `Response` directly so you should check that and send a redirect response instead of attempting to render an HTML document:
+
+```js filename=server.jsx lines=[5-10]
+app.get("*", async (req, res) => {
+  let fetchRequest = createFetchRequest(req);
+  let context = await handler.query(fetchRequest);
+
+  if (
+    context instanceof Response &&
+    [301, 302, 303, 307, 308].includes(context.status)
+  ) {
+    return res.redirect(
+      context.status,
+      context.headers.get("Location")
+    );
+  }
+
+  // Render HTML...
+});
+```
+
+#### Lazy Routes
+
+If you're using [`route.lazy`][lazy] in your routes, then on the client it's possible you have all the data you need to hydrate, but you don't yet have the route definitions! Ideally, your setup would determine the matched routes on the server and deliver their route bundles on the critical path such that you don't use `lazy` on your initially matched routes. However, if this is not the case you'll need to load these routes and update them in place _roprior_ to hydrating to avoid the router falling back to a loading state:
```

Comment:
```suggestion
If you're using [`route.lazy`][lazy] in your routes, then on the client it's possible you have all the data you need to hydrate, but you don't yet have the route definitions! Ideally, your setup would determine the matched routes on the server and deliver their route bundles on the critical path such that you don't use `lazy` on your initially matched routes. However, if this is not the case you'll need to load these routes and update them in place _prior_ to hydrating to avoid the router falling back to a loading state:
```